x/rollapp/keeper/keeper.go

```diff
@@ -70,67 +74,40 @@ func (k Keeper) TriggerRollappGenesisEvent(ctx sdk.Context, rollapp types.Rollap
+		return fmt.Errorf("failed to register denom metadata: %w", err)
+	}
+
 	// Call the mint genesis tokens function
 	if err := k.mintRollappGenesisTokens(ctx, rollapp); err != nil {
-		return err
+		return fmt.Errorf("failed to mint genesis tokens: %w", err)
 	}
 	rollapp.GenesisState.IsGenesisEvent = true
 	k.SetRollapp(ctx, rollapp)
 	return nil
 }
 
-func (k Keeper) mintRollappGenesisTokens(ctx sdk.Context, rollapp types.Rollapp) error {
-	for _, acc := range rollapp.GenesisState.GenesisAccounts {
-		denomTrace := utils.GetForeignDenomTrace(rollapp.ChannelId, acc.Amount.Denom)
+// registerDenomMetadata registers the denom metadata for the IBC token
+func (k Keeper) registerDenomMetadata(ctx sdk.Context, rollapp types.Rollapp) error {
+	for i := range rollapp.TokenMetadata {
+		denomTrace := utils.GetForeignDenomTrace(rollapp.ChannelId, rollapp.TokenMetadata[i].Base)
 		traceHash := denomTrace.Hash()
 		// if the denom trace does not exist, add it
 		if !k.transferKeeper.HasDenomTrace(ctx, traceHash) {
 			k.transferKeeper.SetDenomTrace(ctx, denomTrace)
 		}
 
-		ibcDenom := denomTrace.IBCDenom()
-
-		k.RegisterDenomMetadata(ctx, rollapp, ibcDenom, acc.Amount.Denom)
-
-		coinsToMint := sdk.NewCoins(sdk.NewCoin(ibcDenom, acc.Amount.Amount))
-
-		if err := k.bankKeeper.MintCoins(ctx, types.ModuleName, coinsToMint); err != nil {
-			return err
-		}
-		accAddress, err := sdk.AccAddressFromBech32(acc.Address)
-		if err != nil {
-			return err
-		}
-		if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, accAddress, coinsToMint); err != nil {
-			return fmt.Errorf("failed to send coins to account: %w", err)
-		}
-	}
-	return nil
-}
+		ibcBaseDenom := denomTrace.IBCDenom()
 
-func (k Keeper) RegisterDenomMetadata(ctx sdk.Context, rollapp types.Rollapp, ibcBaseDenom, baseDenom string) {
-	// if the rollapp has no token metadata, return
-	if len(rollapp.TokenMetadata) == 0 {
-		k.Logger(ctx).Info("skipping new IBC token for rollapp with no metadata", "rollappID", rollapp.RollappId, "denom", ibcBaseDenom)
-		return
-	}
-	// if the denom metadata already exists, return
-	if k.bankKeeper.HasDenomMetaData(ctx, ibcBaseDenom) {
-		k.Logger(ctx).Info("denom metadata already registered", "rollappID", rollapp.RollappId, "denom", ibcBaseDenom)
-		return
-	}
-	for i := range rollapp.TokenMetadata {
-		// if the rollapp token base doesn't match the base denom, skip it
-		if rollapp.TokenMetadata[i].Base != baseDenom {
-			continue
-		}
 		// create a new token denom metadata where it's base = ibcDenom,
 		// and the rest of the fields are taken from rollapp.metadata
 		metadata := banktypes.Metadata{
 			Description: "auto-generated metadata for " + ibcBaseDenom + " from rollapp " + rollapp.RollappId,
 			Base:        ibcBaseDenom,
 			DenomUnits:  make([]*banktypes.DenomUnit, len(rollapp.TokenMetadata[i].DenomUnits)),
-			Display:     rollapp.TokenMetadata[i].Display,
+			Display:     ibcBaseDenom,
```

Comment:
@omritoptix Not too sure about this, but this is expected further down the line - denom display should match the 0th unit display

Reply:
no, why is that? quite the opposite

Reply:
without this change, the error I get is `metadata must contain a denomination unit with display denom 'arax'`